### ISSUE TITLE
False positive for `Lint/UselessAssignment` when assigning in branch and block

### ISSUE THREAD

**Issue by Earlopain** [rubocop/rubocop#13395]
## Expected behavior
The following code warns about a unused variable:

```rb
changed = false

if Random.rand > 1
  changed = true
end

[].each do
  changed = true
end

puts changed
```

I expect no offense for this case. The first assignment is meaningful, removing it will change the output from `false` to `nil`. Changed with https://github.com/rubocop/rubocop/pull/13389, not released yet.

## Actual behavior

```
$ bundle exec rubocop test.rb --only Lint/UselessAssignment
Inspecting 1 file
W

Offenses:

test.rb:1:1: W: [Correctable] Lint/UselessAssignment: Useless assignment to variable - changed.
changed = false
^^^^^^^

1 file inspected, 1 offense detected, 1 offense autocorrectable
```

## RuboCop version

```
Commit 8454d933
$ bundle exec rubocop -V
1.67.0 (using Parser 3.3.5.0, rubocop-ast 1.32.3, analyzing as Ruby 2.7, running on ruby 3.3.5) [x86_64-linux]
  - rubocop-performance 1.22.1
  - rubocop-rake 0.6.0
  - rubocop-rspec 3.2.0
```

**Comment by monkeyWzr** [comment]
Same situation with me.

```
def foo
  a = true # RuboCop: Useless assignment to variable - `a`. [Lint/UselessAssignment]
  if something
    a = false
  end
  [].each do |n|
    a = false
  end
  a
end
```

**Comment by koic** [comment]
cc @pCosta99

**Comment by pCosta99** [comment]
I'll take a look throughout the day if no one gets to it first

**Comment by opensourceame** [comment]
I've noticed a similar issue.  This code should produce no warnings:

```ruby
class Test
  attr_accessor :a

  def run
    a += 1 if condition_a
    a += 2 if condition_b
    a += 3 if condition_c
  end
end
```

however I get the `Lint/UselessAssignment` error for the final `if` condition.  More disturbing is what happens when I run `rubocop -a` on the file, it "corrects" to this:

```ruby
class Test
  attr_accessor :a

  def run
    a += 1 if condition_a
    a += 2 if condition_b
    a + 3 if condition_c    # WTF!?
  end
end
```

**Comment by Earlopain** [comment]
@opensourceame can you make a new issue for that? It seems to be something different since I get an offense even for earlier versions already.

**Comment by Earlopain** [comment]
Actually, I think that offense is correct. `a` in `run` doesn't reference you attribute, for that you would need to write it as `self.a`

**Comment by ardavis** [comment]
Here's a minimal example of this false positive that I ran into today:

```ruby
def my_method
  state = true

  # This false could be a conditional that returns false
  if false
    state = false
  end

  # This empty array could have been a query that returned no results, for example
  [].each do
    state = false
  end

  # We want to return the boolean, which defaults to true at the top
  state
end
```

This is clearly silly code, but Rubocop reports:

```
rubocop ".../scratches/scratch_14.rb" --only Lint/UselessAssignment
W

Offenses:

.../scratches/scratch_14.rb:2:3: W: [Correctable] Lint/UselessAssignment: Useless assignment to variable - state.
  state = true
  ^^^^^

1 file inspected, 1 offense detected, 1 offense autocorrectable

```

Running that method without `state = true` at the top will cause the method to return `nil` instead of `true`.

**Comment by navels** [comment]
Just ran into this myself
